Subbus/Classes/Subbus.swift

```diff
@@ -8,141 +8,241 @@
+    static func post<T>(event: T)
+    
+    /// Unsubscribes a listener from further events of the specified type.
+    ///
+    /// Parameters:
+    ///     - id: The id of the subscriber to this event.  This is the same as used for subscribing.
+    ///     - event: The event type to unsubscribe from.
+    static func unsubscribe<I, T>(id: I, event: T.Type)
+    
+    /// Unsubscribes a listener from all further events.
+    ///
+    /// Parameters:
+    ///     - id: The id of the subscriber to this event.  This is the same as used for subscribing.
+    static func unsubscribe<I>(id: I)
+    
+    /// Unsubscribes all listeners of the specified event type.
+    ///
+    /// Parameters:
+    ///     - event: The event type from which all observers of this event will be unsubscribed.
     
-    init(identifier: String, eventName: String, listener: NSObjectProtocol) {
-        self.identifier = identifier
-        self.eventName = eventName
-        self.listener = listener
+    static func unsubscribe<T>(event: T.Type)
+    
+    /// Whether Subbus should log subscription and event posts to the console log for debugging purposes.
+    static var logToConsole: Bool { get set }
+}
+
+/// A convenience extension to add shortened functions to the `SubbusProtocol`.
+extension SubbusProtocol {
+    /// Adds a new subscription to Subbus with the specified listener ID and event type.
+    ///
+    /// If the `id` property is a class and it is not retained (ARC disposes of the object), Subbus will automatically remove any subscriptions for that `id`.
+    ///
+    /// Parameters:
+    ///     - id: The id of the subscriber to this event.  Used for unregistering.  Can be `Any` type, but classes and primitives are recommended.  Arrays, structs, etc are supported, but keep in mind they're passed as values.  If your struct/array/etc changes, this means it's a different identifier, similar to a string or integer.
+    ///     - event: The event type to register the subscriber to.  Normally this is a class with no parent class, but there are special event types available.  See the Subbus docs.
+    ///     - callback: The handler block for when the event is `Post`ed.
+    public static func addSubscription<I, T>(id: I, event: T.Type, callback: @escaping (T) -> Void) {
+        addSubscription(id: id, event: event, replace: false, callback: callback)
     }
 }
 
-//Implementation
-public class Subbus {
-    
+/// Conformance to the `SubbusProtocol`.
+public class Subbus: SubbusProtocol {
     //State
-    static var loggingEnabled: Bool = false
-    private static let shared = Subbus()
-    private var listeners = [ListenerWrapper]() //TODO: Dictionary key of EventName with array of listeners - would be more performant
-    private let eventbusCenter = NotificationCenter()
+    static var logToConsole: Bool = false
+    internal static var subscriptions = [Subscription]()
+    internal static var unhandledEvents = [PersistentEvent]()
```

Comment:
Why do we have unhandled events? Is the intention to cache events until there is an appropriate subscription for them to execute on?

Reply:
Yes sir!  Notice that only persistent events can go into this cache.